okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java

```diff
@@ -0,0 +1,256 @@
+        } catch (ExecutionException e) {
+            getDelegate().cancel();
+            if (e.getCause() instanceof IoRemoteException) {
+                // TODO(rfink): Consider unwrapping the RemoteException at the Retrofit/Feign layer for symmetry, #626
+                throw ((IoRemoteException) e.getCause()).getWrappedException();
+            } else if (e.getCause() instanceof IOException) {
+                throw (IOException) e.getCause();
+            } else {
+                throw new IOException("Failed to execute call", e);
+            }
+        } catch (TimeoutException e) {
+            getDelegate().cancel();
+            throw new IOException("Call timed out after: " + syncCallTimeout.toString(), e);
+        }
+    }
+
+    @Override
+    public void enqueue(Callback callback) {
+        super.enqueue(new Callback() {
+            @Override
+            public void onFailure(Call call, IOException exception) {
+                // Fail call if backoffs are exhausted or if no retry URL can be determined.
+                if (!backoffStrategy.nextBackoff().isPresent()) {
+                    callback.onFailure(call, new IOException("Failed to complete the request due to an "
+                            + "IOException", exception));
+                    return;
+                }
+                Optional<HttpUrl> redirectTo = urls.redirectToNext(request().url());
+                if (!redirectTo.isPresent()) {
+                    callback.onFailure(call, new IOException("Failed to determine valid failover URL"
+                            + "for '" + request().url() + "' and base URLs " + urls.getBaseUrls()));
+                    return;
+                }
+
+                Request redirectedRequest = request().newBuilder()
+                        .url(redirectTo.get())
+                        .build();
+                RemotingOkHttpCall retryCall =
+                        client.newCallWithMutableState(redirectedRequest, backoffStrategy, maxNumRelocations - 1);
+                retryCall.enqueue(callback);
+            }
+
+            @Override
+            public void onResponse(Call call, Response response) throws IOException {
+                // Relay successful responses
+                if (response.code() / 100 == 2) {
+                    callback.onResponse(call, response);
+                    return;
+                }
+
+                // Handle to handle QoS situations: retry, failover, etc.
+                Optional<QosException> qosError = qosHandler.handle(response);
+                if (qosError.isPresent()) {
+                    qosError.get().accept(createQosVisitor(callback, call));
+                    return;
+                }
+
+                // Handle responses that correspond to RemoteExceptions / SerializableErrors
+                Optional<RemoteException> httpError = remoteExceptionHandler.handle(response);
```

Comment:
I noticed both this `RemoteExceptionResponseHandler` and `IoExceptionResponseHandler` check for `Response.isSuccessful`, which will never be true if we're already filtering this out at the top of this method.

Reply:
better safe than sorry?

Reply:
Hmm, no I don't think it's desirable to have a branch that we know will never be hit.  If someone breaks the desired happy path somehow, I'd prefer things to fail as fast as possible (and get fixed), not be obscured by some kind of 'defense in depth' if statements.